vpr/src/place/place_macro.h

```diff
@@ -143,28 +129,138 @@
+     * heads of the placement macros using criteria depending on the type of placement
+     * macros. For carry chains, the heads of the placement macros are blocks with
+     * carry_in's not connected to any nets (OPEN) while the carry_out's connected to the
+     * netlist with only 1 SINK.
+     * (3) Traverse from the heads to the tails of the placement macros and load the
+     * information in the t_pl_macro data structure. Similar to (2), tails are identified
+     * with criteria depending on the type of placement macros. For carry chains, the
+     * tails are blocks with carry_out's not connected to any nets (OPEN) while the
+     * carry_in's is connected to the netlist which has only 1 SINK.
+     * @param directs
+     */
+    void alloc_and_load_placement_macros(const std::vector<t_direct_inf>& directs);
+
+    int get_imacro_from_iblk(ClusterBlockId iblk) const;
+
+    void set_imacro_for_iblk(int imacro, ClusterBlockId blk_id);
+
+    /**
+     * @brief Finds the head block of the macro that contains a given clustered block
+     * @details Placement macro head is the base of the macro, where the locations of the other macro members can be
+     * calculated using base.loc + member.offset.
+     *
+     * @return The Id of a clustered block that is the head of a macro that the given clustered block is part of.
+     */
+    ClusterBlockId macro_head(ClusterBlockId blk) const;
+
+    const std::vector<t_pl_macro>& macros() const;
+
+//    t_pl_macro& operator[](size_t idx);
+    const t_pl_macro&  operator[](int idx) const;
+
+  private:
+
+    /**
+     * @brief This array allow us to quickly find pins that could be in a direct connection.
+     * @details Values stored is the index of the possible direct connection as specified in the arch file,
+     * OPEN (-1) is stored for pins that could not be part of a direct chain connection.
+     * [0...device_ctx.num_block_types-1][0...num_pins-1]
+     */
+    std::vector<std::vector<int>> idirect_from_blk_pin_;
+
+    /**
+     * @brief This array stores the value SOURCE if the pin is the from_pin,
+     * SINK if the pin is the to_pin in the direct connection as specified in the arch file,
+     * OPEN (-1) is stored for pins that could not be part of a direct chain connection.
+     * [0...device_ctx.num_block_types-1][0...num_pins-1]
+     */
+    std::vector<std::vector<int>> direct_type_from_blk_pin_;
+
+    /**
+     * @brief Maps a blk_num to the corresponding macro index.
+     * @details If the block is not part of a macro, the value OPEN (-1) is stored.
+     * [0...cluster_ctx.clb_nlist.blocks().size()-1]
+     */
+    vtr::vector_map<ClusterBlockId, int> imacro_from_iblk_;
+
+    ///@brief Stores all the placement macros (usually carry chains).
+    std::vector<t_pl_macro> pl_macros_;
+
+  private:
+    int find_all_the_macro_(std::vector<ClusterBlockId>& pl_macro_member_blk_num_of_this_blk,
```

Comment:
If you know what they do, write a Doxygen comment.  If not, don't worry about it.